### PR TITLE
Support compilation for x86 as a static library using CMake.

### LINKED ISSUE #176: [Feature Request] Support x86 as compilation target
<a href="https://107-systems.org/"></a>

### ⚡ Feature Request

This library should not be only used in combination with various Arduino cores but should be also compilable as a static C++ library and used with executable compiled for and executed on x86.

### PR DESCRIPTION
**Pull request by aentinger**
This fixes #176.

### CODE REVIEW

**Comment by aentinger** [comment]
Hi @pavel-kirienko :coffee: :wave:

I've got some issue (**invalid use of incomplete type**) with my forward declarations. It works, but the compiler is not happy when I turn up -Wpedantic/-Werror:

```bash
-Arduino-Cyphal/src/Publisher.ipp: In member function ‘bool impl::Publisher<T>::publish(const T&)’:
107-Arduino-Cyphal/src/Publisher.ipp:45:10: error: invalid use of incomplete type ‘class Node’ [-Werror]
   45 |   return _node_hdl.enqueue_transfer(_tx_timeout_usec,
      |          ^~~~~~~~~

107-Arduino-Cyphal/src/Publisher.hpp:21:7: note: forward declaration of ‘class Node’
   21 | class Node;
      |       ^~~~
```

I'm not sure how to best solve this, as both Node and Publisher use the other class. Do you have any suggestion?

**Comment by pavel-kirienko** [comment]
You cannot invoke methods on an incomplete type, so `.enqueue_transfer()` is not valid if `Node` is only forward-declared. I recommend extracting the Publisher base class into a separate header file to break the cycling dependency such that the resulting dependency is like this: PublisherBase <- Node <- Publisher<>

**Comment by aentinger** [comment]
Thank you for the pointer, I managed to break the circular dependency by extracting base classes into a separate header file (and inventing one, where none existed before).

**Comment by aentinger** [comment]
Hi @pavel-kirienko :coffee: :wave:

The purpose here, as you might have guessed is to allow compilation of this library to a static library which can be linked against any C++ executable. Consequently this is a bit like [libcyphal](https://github.com/OpenCyphal-Garage/libcyphal), better because it actually supports OpenCyphal, worse because it only supports CAN/CAN FD. The library shall be named `libcyphalcan++` but I'm open to taking suggestions.

Please also take a look at the example provided and let me know if there's anything missing for a fully functional OpenCyphal node.

Right now this PR is in the development stage, I'll be asking for a formal review later.

**Comment by pavel-kirienko** [comment]
>worse because it only supports CAN/CAN FD. The library shall be named libcyphalcan++ but I'm open to taking suggestions.

Surely it can be extended down the road to also accommodate Cyphal/UDP via libudpard? If so, then maybe "can" is unnecessary in the name.

Why can't it be a header-only library? I'm not even talking about C++ modules since we are stuck with the good old C++11 but at least header-only would remove the need for compilation and simplify integration.

File: src/Publisher.hpp [hunk 1]
```diff
@@ -32,7 +28,7 @@ namespace impl
  **************************************************************************************/
 
 template <typename T>
-class Publisher
+class Publisher : public PublisherBase<T>
```
Comment (pavel-kirienko) [inline src/Publisher.hpp:31]:
```suggestion
class Publisher final : public PublisherBase<T>
```

File: src/ServiceClientBase.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace impl
+{
+
+/**************************************************************************************
+ * CLASS DECLARATION
+ **************************************************************************************/
+
+template <typename T_REQ>
+class ServiceClientBase : public SubscriptionBase
```
Comment (pavel-kirienko) [inline src/ServiceClientBase.hpp:31]: Does this inheritance actually have to be public?

File: src/ServiceServerBase.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ **************************************************************************************/
+
+namespace impl
+{
+
+/**************************************************************************************
+ * CLASS DECLARATION
+ **************************************************************************************/
+
+class ServiceServerBase : public SubscriptionBase
```
Comment (pavel-kirienko) [inline src/ServiceServerBase.hpp:30]: Same question here -- does it have to be public?

File: src/nodeinfo/NodeInfo.cpp [hunk 4]
```diff
@@ -11,6 +11,10 @@
 
 #include "NodeInfo.h"
 
+#undef max
+#undef min
```
Comment (pavel-kirienko) [inline src/nodeinfo/NodeInfo.cpp:15]: Haha, where's this coming from? I have only seen this in WIN API so far.

File: src/nodeinfo/NodeInfo.cpp [hunk 5]
```diff
@@ -39,7 +43,7 @@ NodeInfo::NodeInfo(Node & node_hdl,
 
   memcpy(_node_info.unique_id, unique_id.data(), sizeof(_node_info.unique_id));
 
-  _node_info.name.count = std::min(name.length(), uavcan_node_GetInfo_Response_1_0_name_ARRAY_CAPACITY_);
+  _node_info.name.count = std::min(static_cast<size_t>(name.length()), static_cast<size_t>(uavcan_node_GetInfo_Response_1_0_name_ARRAY_CAPACITY_));
```
Comment (pavel-kirienko) [inline src/nodeinfo/NodeInfo.cpp:46]:
```suggestion
  _node_info.name.count = std::min(name.length(), static_cast<size_t>(uavcan_node_GetInfo_Response_1_0_name_ARRAY_CAPACITY_));
```

redundant cast

File: src/register/RegisterBase.cpp [hunk 6]
```diff
@@ -26,7 +28,7 @@ RegisterBase::RegisterBase(char const * name,
   [name]() -> uavcan_register_Name_1_0
   {
     uavcan_register_Name_1_0 n;
-    n.name.count = std::min(strlen(name), uavcan_register_Name_1_0_name_ARRAY_CAPACITY_);
+    n.name.count = std::min(static_cast<size_t>(strlen(name)), static_cast<size_t>(uavcan_register_Name_1_0_name_ARRAY_CAPACITY_));
```
Comment (pavel-kirienko) [inline src/register/RegisterBase.cpp:31]:
```suggestion
    n.name.count = std::min(std::strlen(name), static_cast<size_t>(uavcan_register_Name_1_0_name_ARRAY_CAPACITY_));
```

redundant cast

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
There are some nitpicks that also apply to the main library. I overlooked them during the earlier reviews.

File: src/nodeinfo/NodeInfo.cpp (hunk 4 above)
Comment (aentinger) [reply src/nodeinfo/NodeInfo.cpp:15]: Ze Arduino API. There's all kind of [weird magic](https://github.com/arduino/ArduinoCore-API/blob/master/api/Common.h#L124-L149) done over there.

File: src/nodeinfo/NodeInfo.cpp (hunk 5 above)
Comment (aentinger) [reply src/nodeinfo/NodeInfo.cpp:46]: Sadly, this cast is not redundant in my build.

File: src/register/RegisterBase.cpp (hunk 6 above)
Comment (aentinger) [reply src/register/RegisterBase.cpp:31]: Sadly, this cast is not redundant in my build.

File: extras/cyphal++/include/SocketCAN.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+{
+public:
+  enum class Protocol : bool
+  {
+    Classic = false,
+    FD = true,
+  };
+
+   SocketCAN(std::string const & iface_name, Protocol const protocol);
+  ~SocketCAN();
```
[the remark this answers is not in the document]
Comment (aentinger) [reply extras/cyphal++/include/SocketCAN.h:33]: Why? No virtual methods here ...

[the remark this answers is not in the document]
Comment (aentinger) [reply extras/cyphal++/include/SocketCAN.h:33]: Fixed in cd7821862cce618073de88229455f77c2ae331f5.

File: src/ServiceServerBase.hpp (hunk 3 above)
Comment (aentinger) [reply src/ServiceServerBase.hpp:30]: If the inheritance is private I won't be able to access `canard_rx_subscription` as required [here](https://github.com/107-systems/107-Arduino-Cyphal/blob/fix-176/src/Node.ipp#L48).

File: extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<a href="https://opencyphal.org/"><img align="right" src="https://raw.githubusercontent.com/107-systems/.github/main/logo/opencyphal.svg" width="25%"></a>
+:floppy_disk: `example-01-opencyphal-basic-node`
+================================================
+* Setup a virtual CAN interface (`vcan0`)
+```bash
+sudo . setup_vcan.sh
```
Comment (pavel-kirienko) [inline extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md:6]:
```suggestion
sudo ./setup_vcan.sh
```

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

File: extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md (hunk 8 above)
Comment (aentinger) [reply extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md:6]: Actually `. setup_vcan.sh` would be enough, due to `[ "$UID" -eq 0 ] || exec sudo bash "$0" "$@"`

File: extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+<a href="https://opencyphal.org/"><img align="right" src="https://raw.githubusercontent.com/107-systems/.github/main/logo/opencyphal.svg" width="25%"></a>
+:floppy_disk: `example-01-opencyphal-basic-node`
+================================================
+* Setup a virtual CAN interface (`vcan0`)
+```bash
+. setup_vcan.sh
```
Comment (pavel-kirienko) [inline extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md:6]:
```suggestion
sudo ./setup_vcan.sh
```

Comment (pavel-kirienko) [reply extras/cyphal++/examples/example-01-opencyphal-basic-node/README.md:6]: This is a script, not a sourceable file, and it requires superuser privileges.

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**